<a href="https://colab.research.google.com/github/hxh00138/ai-science-training-series/blob/main/session3_hw_Xiaohan_Hu.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
%matplotlib inline

import tensorflow as tf

import numpy
import matplotlib.pyplot as plt
import time

In [ ]:
(x_train, y_train), (x_test, y_test) = tf.keras.datasets.cifar10.load_data()
x_train = x_train.astype(numpy.float32)
x_test  = x_test.astype(numpy.float32)

x_train /= 255.
x_test  /= 255.

y_train = y_train.astype(numpy.int32)
y_test  = y_test.astype(numpy.int32)

170498071/170498071 [==============================] - 4s 0us/step


In [ ]:
class CIFAR10Classifier(tf.keras.models.Model):

    def __init__(self, activation=tf.nn.tanh):
        tf.keras.models.Model.__init__(self)

        self.conv_1 = tf.keras.layers.Conv2D(32, [3, 3], activation='relu')
        self.conv_2 = tf.keras.layers.Conv2D(64, [3, 3], activation='relu')
        self.pool_3 = tf.keras.layers.MaxPooling2D(pool_size=(2, 2))
        self.drop_4 = tf.keras.layers.Dropout(0.25)
        self.conv_5 = tf.keras.layers.Conv2D(128, [3, 3], activation='relu')
        self.pool_6 = tf.keras.layers.MaxPooling2D(pool_size=(2, 2))
        self.dense_7 = tf.keras.layers.Dense(128, activation='relu')
        self.drop_8 = tf.keras.layers.Dropout(0.5)
        self.dense_9 = tf.keras.layers.Dense(10, activation='softmax')

    def call(self, inputs):

        x = self.conv_1(inputs)
        x = self.conv_2(x)
        x = self.pool_3(x)
        x = self.drop_4(x)
        x = self.conv_5(x)
        x = self.pool_6(x)
        x = tf.keras.layers.Flatten()(x)
        x = self.dense_7(x)
        x = self.drop_8(x)
        x = self.dense_9(x)

        return x

In [ ]:
def train_network_concise(_batch_size, _n_training_epochs, _lr):

    cnn_model = CIFAR10Classifier()

    cnn_model.compile(loss="sparse_categorical_crossentropy", optimizer="adam", metrics=['accuracy'])
    
    history = cnn_model.fit(x_train, y_train, batch_size=_batch_size, epochs=_n_training_epochs)
    return history, cnn_model

In [ ]:
# This took 43 seconds per epoch on my laptop
batch_size = 64
epochs = 5
lr = .01
history, cnn_model = train_network_concise(batch_size, epochs, lr)

Epoch 1/5
782/782 [==============================] - 263s 335ms/step - loss: 1.6956 - accuracy: 0.3785
Epoch 2/5
782/782 [==============================] - 260s 332ms/step - loss: 1.3205 - accuracy: 0.5265
Epoch 3/5
782/782 [==============================] - 258s 330ms/step - loss: 1.1604 - accuracy: 0.5889
Epoch 4/5
782/782 [==============================] - 254s 324ms/step - loss: 1.0470 - accuracy: 0.6333
Epoch 5/5
782/782 [==============================] - 253s 324ms/step - loss: 0.9618 - accuracy: 0.6615


#By adding one more convolutional layer and pooling layer and also increasing the epochs, we can increase the accuracy from 50% to 66%.